src/Workspaces/Core/Portable/Workspace/ProjectSystem/ProjectSystemProjectFactory.cs

```diff
@@ -321,26 +356,69 @@ internal void RemoveProjectFromTrackingMaps_NoLock(ProjectId projectId)
 
         var project = Workspace.CurrentSolution.GetRequiredProject(projectId);
 
-        if (_projectReferenceInfoMap.TryGetValue(projectId, out var projectReferenceInfo))
+        ApplyBatchChangeToWorkspace_NoLock((solutionChanges, projectUpdateState) =>
         {
-            // If we still had any output paths, we'll want to remove them to cause conversion back to metadata references.
-            // The call below implicitly is modifying the collection we've fetched, so we'll make a copy.
-            var solutionChanges = new SolutionChangeAccumulator(Workspace.CurrentSolution);
+            var project = Workspace.CurrentSolution.GetRequiredProject(projectId);
 
-            foreach (var outputPath in projectReferenceInfo.OutputPaths.ToList())
+            if (projectUpdateState.ProjectReferenceInfos.TryGetValue(projectId, out var projectReferenceInfo))
             {
-                RemoveProjectOutputPath_NoLock(solutionChanges, projectId, outputPath);
+                // If we still had any output paths, we'll want to remove them to cause conversion back to metadata references.
+                // The call below implicitly is modifying the collection we've fetched, so we'll make a copy.
+                foreach (var outputPath in projectReferenceInfo.OutputPaths.ToList())
+                {
+                    projectUpdateState = RemoveProjectOutputPath_NoLock(solutionChanges, projectId, outputPath, projectUpdateState, SolutionClosing, SolutionServices);
+                }
+
+                projectUpdateState = projectUpdateState with
+                {
+                    ProjectReferenceInfos = projectUpdateState.ProjectReferenceInfos.Remove(projectId)
+                };
             }
 
-            ApplyBatchChangeToWorkspace_NoLock(solutionChanges);
+            return projectUpdateState;
+        }, onAfterUpdate: (projectUpdateState) =>
+        {
+            ImmutableInterlocked.TryRemove<ProjectId, string?>(ref _projectToMaxSupportedLangVersionMap, projectId, out _);
+            ImmutableInterlocked.TryRemove(ref _projectToDependencyNodeTargetIdentifier, projectId, out _);
 
-            _projectReferenceInfoMap.Remove(projectId);
-        }
+            _onProjectRemoved?.Invoke(project);
+        });
```

Comment:
i wouldn't mind if the onAfterUpdates had some explanatory text that they are only called once total success has happened updating the real workspace solution.